src/base/utils/os.cpp

```diff
@@ -0,0 +1,412 @@
+            qDebug("Python versions nb: %d", versions.size());
+            versions.sort();
+
+            bool found = false;
+            while (!found && !versions.empty()) {
+                const QString version = versions.takeLast() + "\\InstallPath";
+                LPWSTR lpSubkey = new WCHAR[version.size() + 1];
+                version.toWCharArray(lpSubkey);
+                lpSubkey[version.size()] = 0;
+
+                HKEY hkInstallPath;
+                res = ::RegOpenKeyExW(hkPythonCore, lpSubkey, 0, samDesired, &hkInstallPath);
+                delete[] lpSubkey;
+
+                if (res == ERROR_SUCCESS) {
+                    qDebug("Detected possible Python v%s location", qUtf8Printable(version));
+                    path = getRegValue(hkInstallPath);
+                    ::RegCloseKey(hkInstallPath);
+
+                    if (!path.isEmpty() && QDir(path).exists("python.exe")) {
+                        qDebug("Found python.exe at %s", qUtf8Printable(path));
+                        found = true;
+                    }
+                }
+            }
+
+            if (!found)
+                path = QString();
+
+            ::RegCloseKey(hkPythonCore);
+        }
+
+        return path;
+    }
+
+    QString getAssoc(const QString &cls)
+    {
+        return (QSettings {"HKEY_CLASSES_ROOT", QSettings::NativeFormat}).value(cls + "/Default").toString();
+    }
+
+    void setAssoc(const QString &cls, const QString &value)
+    {
+        QSettings reg {"HKEY_CURRENT_USER\\Software\\Classes", QSettings::NativeFormat};
+        const QString oldProgID = reg.value(cls + "/Default").toString();
+        if (!oldProgID.isEmpty() && (oldProgID != value))
+            reg.setValue(cls + "/OpenWithProgids/" + oldProgID, "");
+        reg.setValue(cls + "/Default", value);
+        ::SHChangeNotify(SHCNE_ASSOCCHANGED, SHCNF_IDLIST, 0, 0);
+    }
+}
+
+QString Utils::OS::getPythonPath()
+{
+    QString path = pythonSearchReg(USER);
+    if (!path.isEmpty())
+        return path;
+
+    path = pythonSearchReg(SYSTEM_32BIT);
+    if (!path.isEmpty())
+        return path;
+
+    path = pythonSearchReg(SYSTEM_64BIT);
+    if (!path.isEmpty())
+        return path;
+
+    // Fallback: Detect python from default locations
+    const QStringList dirs = QDir("C:/").entryList(QStringList("Python*"), QDir::Dirs, QDir::Name | QDir::Reversed);
+    foreach (const QString &dir, dirs) {
+        const QString path("C:/" + dir + "/");
+        if (QFile::exists(path + "python.exe"))
+            return path;
+    }
+
+    return QString();
+}
+
+bool Utils::OS::areAssociationsSet() noexcept
+{
+#ifndef WIN_HAS_DEFAULT_PROGRAMS_API
+    Q_ASSERT(QSysInfo::WindowsVersion < QSysInfo::WV_VISTA);
+    return (isTorrentFileAssocSet() && isMagnetLinkAssocSet());
+#else
+    Q_ASSERT(QSysInfo::WindowsVersion < QSysInfo::WV_WINDOWS8);
+
+    if (QSysInfo::WindowsVersion < QSysInfo::WV_VISTA)
```

Comment:
Why does the condition repeat the assert in line 256?

Reply:
That's right.